common/variant/src/main/java/org/apache/spark/types/variant/VariantBuilder.java

```diff
@@ -214,6 +214,22 @@ public void appendTimestampNtz(long microsSinceEpoch) {
     writePos += 8;
   }
 
+  public void appendYearMonthInterval(long value, byte startField, byte endField) {
+    checkCapacity(1 + 5);
+    writeBuffer[writePos++] = primitiveHeader(YEAR_MONTH_INTERVAL);
+    writeBuffer[writePos++] = (byte) (startField | (endField << 1));
+    writeLong(writeBuffer, writePos, value, 4);
+    writePos += 4;
+  }
+
+  public void appendDayTimeInterval(long value, byte startField, byte endField) {
+    checkCapacity(1 + 9);
+    writeBuffer[writePos++] = primitiveHeader(DAY_TIME_INTERVAL);
+    writeBuffer[writePos++] = (byte) (startField | (endField << 2));
```

Comment:
Should we mask `startField` and `endField` to make sure we are only using the desired bits?

Or, we could do checks on the inputs?